File: frontend/src/services/appointments.ts
import type {
  AppointmentFilterReferenceData,
  AppointmentQuery,
  AppointmentReferenceData,
  CreateAppointmentPayload,
  CreateAppointmentResponse,
  PaginatedAppointmentsResponse,
  RescheduleAppointmentPayload,
  RescheduleAppointmentResponse,
  UpdateAppointmentPayload,
  UpdateAppointmentResponse,
} from "../types/appointments";

/* ----------------------------------------------------- */
/* Existing API model                                    */
/* ----------------------------------------------------- */

export interface AppointmentApiModel {
  appt_id: string;
  appt_date: string;
  customer_name: string | null;
  customer_id: string | null;
  facility_name: string | null;
  facility_id: string | null;
  scheduled_time: string;
  carrier_name: string | null;
  status: string | null;
}

/* ----------------------------------------------------- */
/* API configuration                                     */
/* ----------------------------------------------------- */

const API_BASE_URL = "http://127.0.0.1:8000";

/* ----------------------------------------------------- */
/* Existing endpoint                                     */
/* ----------------------------------------------------- */

export async function getAppointments(): Promise<
  AppointmentApiModel[]
> {
  const response = await fetch(
    `${API_BASE_URL}/api/appointments`,
  );

  if (!response.ok) {
    throw new Error(
      `Unable to load appointments: ${response.status}`,
    );
  }

  return response.json();
}

/* ----------------------------------------------------- */
/* Paginated endpoint                                    */
/* ----------------------------------------------------- */

export async function getPaginatedAppointments(
  query: AppointmentQuery,
): Promise<PaginatedAppointmentsResponse> {
  const params = new URLSearchParams({
    page: String(query.page),
    page_size: String(query.pageSize),
  });

  if (query.facilityId) {
    params.set("facility_id", query.facilityId);
  }

  if (query.customerId) {
    params.set("customer_id", query.customerId);
  }

  if (query.carrierId) {
    params.set("carrier_id", query.carrierId);
  }

  if (query.dockId) {
    params.set("assigned_dock_id", query.dockId);
  }

  if (query.appointmentType) {
    params.set("appointment_type", query.appointmentType);
  }

  if (query.dateFrom) {
    params.set("date_from", query.dateFrom.split("T")[0]);
  }

  if (query.dateTo) {
    params.set("date_to", query.dateTo.split("T")[0]);
  }

  if (query.palletMin !== undefined) {
    params.set("pallet_min", String(query.palletMin));
  }

  if (query.palletMax !== undefined) {
    params.set("pallet_max", String(query.palletMax));
  }

  if (query.skuMin !== undefined) {
    params.set("sku_min", String(query.skuMin));
  }

  if (query.skuMax !== undefined) {
    params.set("sku_max", String(query.skuMax));
  }

  if (query.status) {
    params.set("status", query.status);
  }

  if (query.riskLevel) {
    params.set("risk_level", query.riskLevel);
  }

  if (query.outcome) {
    params.set("outcome", query.outcome);
  }

  if (query.search?.trim()) {
    params.set("search", query.search.trim());
  }

  if (query.sortBy && query.sortDirection) {
    params.set("sort_by", query.sortBy);
    params.set("sort_direction", query.sortDirection);
  }

  const response = await fetch(
    `${API_BASE_URL}/api/appointments/paged?${params.toString()}`,
  );

  if (!response.ok) {
    throw new Error(
      `Unable to load appointments: ${response.status}`,
    );
  }

  return response.json();
}

export async function getAppointmentReferenceData(): Promise<AppointmentReferenceData> {
  const response = await fetch(`${API_BASE_URL}/api/appointments/reference-data/options`);
  if (!response.ok) {
    throw new Error(`Unable to load appointment options: ${response.status}`);
  }
  return response.json();
}

export async function getAppointmentFilterOptions(
  query: Pick<
    AppointmentQuery,
    | "facilityId"
    | "customerId"
    | "carrierId"
    | "appointmentType"
    | "dateFrom"
    | "dateTo"
  >,
  signal?: AbortSignal,
): Promise<AppointmentFilterReferenceData> {
  const params = new URLSearchParams();
  if (query.facilityId) params.set("facility_id", query.facilityId);
  if (query.customerId) params.set("customer_id", query.customerId);
  if (query.carrierId) params.set("carrier_id", query.carrierId);
  if (query.appointmentType) params.set("appointment_type", query.appointmentType);
  if (query.dateFrom) params.set("date_from", query.dateFrom.split("T")[0]);
  if (query.dateTo) params.set("date_to", query.dateTo.split("T")[0]);

  const response = await fetch(
    `${API_BASE_URL}/api/appointments/reference-data/filter-options?${params.toString()}`,
    { signal },
  );
  if (!response.ok) {
    throw new Error(`Unable to load cascading filter options: ${response.status}`);
  }

  const payload = await response.json();
  return {
    facilities: payload.facilities ?? [],
    customers: payload.customers ?? [],
    carriers: payload.carriers ?? [],
    appointmentTypes: payload.appointment_types ?? [],
  };
}

export async function createAppointment(
  payload: CreateAppointmentPayload,
): Promise<CreateAppointmentResponse> {
  const response = await fetch(`${API_BASE_URL}/api/appointments`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const rawBody = await response.text();
    let message = `Unable to create appointment: ${response.status}`;

    if (rawBody) {
      try {
        const body = JSON.parse(rawBody);

        if (typeof body?.message === "string") {
          message = body.message;
        } else if (typeof body?.detail === "string") {
          message = body.detail;
        } else if (Array.isArray(body?.detail)) {
          message = body.detail
            .map((item: { loc?: unknown[]; msg?: string }) => {
              const location = item.loc?.slice(1).join(".") || "request";
              return `${location}: ${item.msg ?? "Invalid value"}`;
            })
            .join("; ");
        }
      } catch {
        message = rawBody;
      }
    }

    throw new Error(message);
  }

  return response.json();
}


export async function updateAppointment(
  appointmentId: string,
  payload: UpdateAppointmentPayload,
): Promise<UpdateAppointmentResponse> {
  const response = await fetch(`${API_BASE_URL}/api/appointments/${appointmentId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const rawBody = await response.text();
    let message = `Unable to update appointment: ${response.status}`;
    if (rawBody) {
      try {
        const body = JSON.parse(rawBody);
        message = body?.message ?? body?.detail ?? message;
      } catch {
        message = rawBody;
      }
    }
    throw new Error(message);
  }
  return response.json();
}


export async function rescheduleAppointment(
  appointmentId: string,
  payload: RescheduleAppointmentPayload,
): Promise<RescheduleAppointmentResponse> {
  const response = await fetch(
    `${API_BASE_URL}/api/appointments/${appointmentId}/reschedule`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );

  if (!response.ok) {
    const rawBody = await response.text();
    let message = `Unable to reschedule appointment: ${response.status}`;
    if (rawBody) {
      try {
        const body = JSON.parse(rawBody);
        message = body?.message ?? body?.detail ?? message;
      } catch {
        message = rawBody;
      }
    }
    throw new Error(message);
  }

  return response.json();
}

File: backend/app/api/appointments.py
from datetime import date
from typing import Any

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Appointment
from app.repositories.appointment_repository import (
    AppointmentRepository,
)
from app.schemas import (
    AppointmentCreate,
    AppointmentCreatedResponse,
    AppointmentReferenceData,
    AppointmentResponse,
    AppointmentUpdate,
    AppointmentUpdatedResponse,
    WhatIfRequest,
    WhatIfResponse,
)
from app.services.appointment_service import (
    AppointmentService,
)


router = APIRouter(
    prefix="/api/appointments",
    tags=["Appointments"],
)


def get_appointment_service(
    db: Session = Depends(get_db),
) -> AppointmentService:
    repository = AppointmentRepository(db)

    return AppointmentService(repository)


@router.get(
    "",
    response_model=list[AppointmentResponse],
)
def get_appointments(
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> list[Appointment]:
    return service.get_all()


@router.get("/paged")
def get_paginated_appointments(
    page: int = Query(
        default=1,
        ge=1,
    ),
    page_size: int = Query(
        default=10,
        ge=10,
        le=50,
    ),
    facility_id: str | None = Query(
        default=None,
    ),
    customer_id: str | None = Query(default=None),
    carrier_id: str | None = Query(default=None),
    appointment_type: str | None = Query(
        default=None,
        pattern="^(Inbound|Outbound)$",
    ),
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    status: str | None = Query(
        default=None,
    ),
    risk_level: str | None = Query(
        default=None,
    ),
    outcome: str | None = Query(
        default=None,
    ),
    search: str | None = Query(
        default=None,
    ),
    sort_by: str | None = Query(
        default=None,
    ),
    sort_direction: str | None = Query(
        default=None,
        pattern="^(asc|desc)$",
    ),
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> dict[str, Any]:
    allowed_page_sizes = {
        10,
        20,
        30,
        40,
        50,
    }

    if page_size not in allowed_page_sizes:
        page_size = 10

    return service.get_paginated(
        page=page,
        page_size=page_size,
        facility_id=facility_id,
        customer_id=customer_id,
        carrier_id=carrier_id,
        appointment_type=appointment_type,
        date_from=date_from,
        date_to=date_to,
        status=status,
        risk_level=risk_level,
        outcome=outcome,
        search=search,
        sort_by=sort_by,
        sort_direction=sort_direction,
    )


@router.get(
    "/reference-data/options",
    response_model=AppointmentReferenceData,
)
def get_appointment_reference_data(
    service: AppointmentService = Depends(get_appointment_service),
) -> dict[str, Any]:
    return service.get_reference_data()

@router.get("/reference-data/filter-options")
def get_appointment_filter_options(
    facility_id: str | None = Query(default=None),
    customer_id: str | None = Query(default=None),
    carrier_id: str | None = Query(default=None),
    appointment_type: str | None = Query(
        default=None,
        pattern="^(Inbound|Outbound)$",
    ),
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    service: AppointmentService = Depends(get_appointment_service),
) -> dict[str, Any]:
    return service.get_filter_reference_data(
        facility_id=facility_id,
        customer_id=customer_id,
        carrier_id=carrier_id,
        appointment_type=appointment_type,
        date_from=date_from,
        date_to=date_to,
    )


@router.get("/{appt_id}/details")
def get_appointment_details(
    appt_id: str,
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> dict[str, Any]:
    return service.get_details(appt_id)


@router.post(
    "/{appt_id}/what-if",
    response_model=WhatIfResponse,
)
def run_appointment_what_if(
    appt_id: str,
    payload: WhatIfRequest,
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> WhatIfResponse:
    return service.run_what_if(
        appt_id=appt_id,
        payload=payload,
    )


@router.get(
    "/{appt_id}",
    response_model=AppointmentResponse,
)
def get_appointment(
    appt_id: str,
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> Appointment:
    return service.get_by_id(appt_id)


@router.patch(
    "/{appt_id}",
    response_model=AppointmentUpdatedResponse,
)
def update_appointment(
    appt_id: str,
    payload: AppointmentUpdate,
    service: AppointmentService = Depends(get_appointment_service),
) -> dict[str, Any]:
    return service.update(appt_id, payload)


@router.post(
    "",
    response_model=AppointmentCreatedResponse,
    status_code=201,
)
def create_appointment(
    payload: AppointmentCreate,
    service: AppointmentService = Depends(
        get_appointment_service
    ),
) -> dict[str, Any]:
    return service.create(payload)